### PR TITLE
Add ET Releases to Collector models

### PR DESCRIPTION
**Pull request by jasinner**
The main changes here are the 2nd and 3rd commits. The first of those commits adds the ET Release data to the Collector models. The other one persists most of the additional ET data not stored into the primary ET Collector model fields into a new meta_attr JSONField attribute.

This will be useful for the data migration in CORGI-843 and CORGI-737 for finding CPEs for some streams.

I also deleted one environment variant converted it to a test constant to make the developer setup experience slightly simpler.

This feature is blocked by CORGI-867 so I added a commit to fix that also.

### CODE REVIEW

**Comment by jasinner** [comment]
@RedHatProductSecurity/corgi-devs please review

**Comment by jasinner** [comment]
I ran the load_variants() function in stage and it successfully fixed the issue described in CORGI-867

File: tox.ini
Comment (juspence) [inline tox.ini:15]: I haven't looked in a while but I think this var might also be defined in Gitlab CI variables, and maybe other places too. Let's make sure we fix those as well.

File: corgi/collectors/models.py
Comment (juspence) [inline corgi/collectors/models.py:38]: Sorry for the dumb questions, I don't know much about these. But could you briefly explain what a Release in ET actually is and how it gets used?

I also see both are linked to ET Product Versions - is it possible for one release to include many variants, and for one variant to be included in multiple different releases? Or are they completely unrelated?

File: corgi/collectors/errata_tool.py
Comment (juspence) [inline corgi/collectors/errata_tool.py:158]: LGTM, thanks for fixing this! Let's switch all the other update_or_create usages to use the unique name field instead of the ET ID.

I also wonder if we should delete and recreate all the Collector models in Corgi each day. Or maybe we need to handle models which are removed / renamed but keep the same ID / deleted and recreated in ET with the same name but different IDs / etc. in some other way.

Let's also add more tests for some of this logic.

File: corgi/collectors/errata_tool.py
Comment (juspence) [inline corgi/collectors/errata_tool.py:144]: Should this raise an error instead of just logging a warning?

File: corgi/collectors/errata_tool.py
Comment (juspence) [inline corgi/collectors/errata_tool.py:114]: We should only use one field which is unique=True in update_or_create keyword arguments, or all the fields that appear together in a UniqueConstraint.

So I think name alone is enough and et_id can be part of the defaults like other methods. I know both et_id and name are technically unique, but they don't appear in a constraint together so I'm not sure if using both will cause some subtle issue.

File: corgi/collectors/errata_tool.py
Comment (juspence) [inline corgi/collectors/errata_tool.py:84]: Did you forget to add `"meta_attr": product_version` here, or was it left off on purpose? For example, if it only has two keys "id" and "name" that we already store elsewhere?

File: corgi/collectors/errata_tool.py
Comment (juspence) [inline corgi/collectors/errata_tool.py:103]: This will work, and I think this is safe if we only call this collector logic from a singleton Celery task once a day. But if any other code updates or manipulates the Collector models, let's put this inside a transaction.

Maybe it's even better to delete all the Collector models each day before we recreate them, so that we can clean up stale data that was removed from ET, and also handle updates in a slightly simpler way. Or maybe that will cause problems elsewhere, I'm not sure what links to these models or how they're used.

File: corgi/collectors/errata_tool.py
Comment (juspence) [inline corgi/collectors/errata_tool.py:111]: Let's keep this relationships key a part of the attributes dict instead of popping it, but still pop off relationships["brew_tags"]. That way, other non-Brew tag relationships (if any) will still get saved into the meta_attr instead of being lost.

File: corgi/collectors/errata_tool.py
Comment (juspence) [inline corgi/collectors/errata_tool.py:93]: Not related to this PR, but we fiddle with Brew tag suffixes in another part of the code too, I think...maybe the prod-defs collector?

Do we need to combine all that logic in one place, or should we strip even more suffixes here, like "-container-released" or whatever?

**Review by juspence — APPROVED** [review]
I think this looks good, but let's add a few more tests to make sure we handle e.g. models being updated (or recreated with different names or IDs) in ET.

File: corgi/collectors/errata_tool.py
Comment (jasinner) [reply corgi/collectors/errata_tool.py:84]: No, there are no additional details here, as these are just the product_versions from the product relationshipts. In the load_product_version function we set the meta_attr. In this function we save the product_versions and their relationships to products, then in load_product_versions we set the meta_attr and brew_tags which are not present in the ET products API endpoint.

File: corgi/collectors/errata_tool.py
Comment (jasinner) [reply corgi/collectors/errata_tool.py:103]: Good idea, I will update it to recreate the collector models each time.

File: corgi/collectors/errata_tool.py
Comment (jasinner) [reply corgi/collectors/errata_tool.py:93]: Perhaps just creating a constant for "-candidate" will give a maintainer a clue that the 2 functions both adjust the raw values? I think it makes sense to adjust the raw values in both function as we currently do. One is brew_tag data coming from ET, another is brew_tag data coming from prod_defs. 

You did make me realize that we also should probably remove the '-candidate' suffix from the new load_et_releases function as well, so that brew tag matching can behave the same between product models and collector models. I'll adjust that.

Brew tags in ET have '-candidate' tags because Engineers never attached released builds to an Errata. ET switches the brew tags to '-released' once they are attached to an Erratum and released. That's why the ET data and prod_defs data differ. Some products don't use the brew tag candidate and released suffixes at all.

File: corgi/collectors/models.py
Comment (jasinner) [reply corgi/collectors/models.py:38]: To be honest I don't know much about Releases. I don't know what their relationship to variants are, if any. Probably they are only related via product versions. I do know that:

- They can be used in a errata search endpoint as a way of grouping Errata
- They declare relationships on both product versions and brew_tags
- As described in CORGI-737 they can be used to override brew_tags defined in product versions when attached builds to an erratum.

File: tox.ini
Comment (jasinner) [reply tox.ini:15]: Yup, good idea. I'll remove the variable from Gitlab CI after we merge this.

File: corgi/collectors/errata_tool.py
Comment (jasinner) [reply corgi/collectors/errata_tool.py:158]: I'll try deleting and recreating the models.

File: corgi/collectors/errata_tool.py
Comment (jasinner) [reply corgi/collectors/errata_tool.py:144]: There are many of these warning logs for each run of the load_et_productions task today. Hopefully there will be less once we delete/recreate the models. If that's the case raising an error would make more sense.

File: corgi/collectors/errata_tool.py
Comment (jasinner) [reply corgi/collectors/errata_tool.py:111]: ok

File: corgi/collectors/errata_tool.py
Comment (jasinner) [reply corgi/collectors/errata_tool.py:114]: I updated the whole function to recreate the objects, so we no longer use update_or_create here.

**Comment by jasinner** [comment]
I got some integrity errors when recreating the collector models, so I pushed to a different branch 'et-models-recreate'. I might abandon that approach.

**Comment by juspence** [comment]
> I got some integrity errors when recreating the collector models, so I pushed to a different branch 'et-models-recreate'. I might abandon that approach.

I think it's fine to use the current approach if deleting and recreating doesn't work. We can look into it further in a future PR, it doesn't have to be done here if it's difficult or causing problems.

File: corgi/collectors/models.py
Comment (juspence) [reply corgi/collectors/models.py:38]: Ack, current code is fine as-is I think. I just wanted to check if we should add a link from Releass to Variants, so I think the current code is fine. We can always add those links in the future if needed later.

File: tox.ini
Comment (juspence) [reply tox.ini:15]: I just checked again and I see the .gitlab-ci.yml pipeline definition also uses this value. Once you remove that and merge this, I think there should be no other usages and deleting the variable in the remote Gitlab CI variables should be safe. Thanks!

**Comment by jasinner** [comment]
Closed in favour of pull/649

**Comment by jasinner** [comment]
> I think this looks good, but let's add a few more tests to make sure we handle e.g. models being updated (or recreated with different names or IDs) in ET.

I added a bunch of test for the new slow_load_stream_errata task.

**Pull request closed without merging**
